module/Database/src/Service/Payment.php

```diff
@@ -0,0 +1,356 @@
+    {
+        $lastCheckoutStub = $this->checkoutSessionMapper->findLatest($prospectiveMember);
+
+        if (null === $lastCheckoutStub) {
+            // We have a problem, a prospective member must have at least one checkout session associated with it.
+            // Ignore it for now and create a new Checkout Session.
+            return $this->getCheckoutLink($prospectiveMember);
+        }
+
+        // We have at least one known Checkout Session on file.
+        if (
+            CheckoutSessionStates::Paid === $lastCheckoutStub->getState()
+            || CheckoutSessionStates::Pending === $lastCheckoutStub->getState()
+        ) {
+            // Checkout Session is finalised or will be after payment processing. Do not allow the prospective member to
+            // do something else.
+            return null;
+        }
+
+        if (CheckoutSessionStates::Failed === $lastCheckoutStub->getState()) {
+            // Last payment failed, so we need to create a new Checkout Session for the user to be able to try again.
+            return $this->getCheckoutLink($prospectiveMember);
+        }
+
+        if (CheckoutSessionStates::Expired === $lastCheckoutStub->getState()) {
+            // The Checkout Session has already been abandoned.
+
+            if ((new DateTime())->add(new DateInterval('PT5M')) >= $lastCheckoutStub->getExpiration()) {
+                // The Checkout Session is completely abandoned, as the maximum expiration for the recovery URL of 30
+                // days has passed (or will pass in 5 minutes). As such, we want to create a new Checkout Session for
+                // this prospective member.
+                return $this->getCheckoutLink($prospectiveMember);
+            }
+
+            // The Checkout Session is not completely dead yet, so return the recovery URL.
+            return $lastCheckoutStub->getRecoveryUrl();
+        }
+
+        // Checkout Session is still valid (at this point, not necessarily when the prospective member finally submits).
+        // We try to retrieve the actual Checkout Session, if this succeeds we return the URL. If not we fail and let
+        // the prospective member know.
+        return $this->getCheckoutSession($lastCheckoutStub->getCheckoutId())?->url;
+    }
+
+    /**
+     * Create a Checkout Session through the Stripe API. The Checkout Session has some required parameters, for more
+     * information on their details see {@link https://stripe.com/docs/api/checkout/sessions/object}.
+     */
+    private function createCheckoutSession(ProspectiveMemberModel $prospectiveMember): ?CheckoutSession
+    {
+        try {
+            return $this->getClient()->checkout->sessions->create([
+                'line_items' => [
+                    [
+                        'price' => $this->config['membership_price_id'],
+                        'quantity' => 1,
+                    ],
+                ],
+                'mode' => 'payment',
```

Comment:
Perhaps we can include a reference that is shown on the members bank or credit card statement? 

```suggestion
                'mode' => 'payment',
                'payment_intent_data' => [
                    'statement_descriptor' => 'GEWIS Membership ' . $prospectiveMember->getLidnr(),
                 ],
```

Reply:
Yes, the only "problem" is that we do not have a link from the prospective lidnr to the actual lidnr.

Reply:
25b64bd